h2o-core/src/main/java/water/udf/DkvClassLoader.java

```diff
@@ -28,20 +38,138 @@ public DkvClassLoader(String jarKeyName, ClassLoader parent) {
   }
   
   public DkvClassLoader(Key jarKey, ClassLoader parent) {
-    super(new URL[] {}, parent);
-    Value v = water.DKV.get(jarKey);
-    // Get local persistent layer and use it to save
-    // content of K/V
-    PersistFS persistFS = (PersistFS) H2O.getPM().getIce();
-    File f = persistFS.getFile(v);
-    // Optimistic delete
-    f.deleteOnExit();
+    super(parent);
+    this.jarKey = jarKey;
+    this.jarCache = buildJarCache(jarKey);
+  }
+  
+  @Override
+  protected Class<?> findClass(String name) throws ClassNotFoundException {
     try {
-      Log.debug("DkvClassLoader: saving " + v + " into " + f.getAbsolutePath());
-      H2O.getPM().store(Value.ICE, v);
-      addURL(f.toURI().toURL());
-    } catch (Exception e) {
-      throw new IllegalArgumentException(e);
+      return super.findClass(name);
+    } catch (ClassNotFoundException e) {
+      // Parent does not contain the requested class, look into cache we built.
+      String path = name.replace('.', '/').concat(".class");
+      byte[] klazzBytes = jarCache.get(path);
+      if (klazzBytes != null && klazzBytes.length > 0) {
+        return defineClass(name, klazzBytes, 0, klazzBytes.length);
+      }
+
+      throw new ClassNotFoundException(name);
+    }
+  }
+  @Override
+  protected URL findResource(String name) {
+    return url(name);
```

Comment:
Should we try to find the resource using the parent (similar to findClass)?

Reply:
That's what `getResource` is doing, `findResource` finds resource in current ClassLoader only. `getResoource` is using internally `findResource` (see that findResource is protected)

Reply:
Got it, thx!